lua/entities/sent_spawnpoint/cl_init.lua

```diff
@@ -1,8 +1,202 @@
+local MESSAGE_COLOR_POINT_SPAWN_COOLDOWN = Color( 255, 0, 0 )
+
+local MESSAGE_TEXT_PLAYER_SPAWN_COOLDOWN = "CANNOT LINK\nCHARGING"
+local MESSAGE_COLOR_PLAYER_SPAWN_COOLDOWN = Color( 255, 0, 0 )
+
+local MESSAGE_TEXT_LINKHINT = "PRESS %s"
+local MESSAGE_COLOR_LINKHINT = Color( 250, 255, 0 )
+
+local MESSAGE_ZOFFSET_HEALTH = -1 * MESSAGE_FONT_SIZE
+local MESSAGE_COLOR_HEALTH = Color( 255, 255, 255 )
+local MESSAGE_COLOR_HEALTH_BAR_FRIENDLY = Color( 0, 180, 255 )
+local MESSAGE_COLOR_HEALTH_BAR_BG_FRIENDLY = Color( 0, 50, 70 )
+local MESSAGE_COLOR_HEALTH_BAR_ENEMY = Color( 255, 0, 0 )
+local MESSAGE_COLOR_HEALTH_BAR_BG_ENEMY = Color( 70, 0, 0 )
+local MESSAGE_BAR_WIDTH = 100 * MESSAGE_CRISPNESS
+local MESSAGE_BAR_HEIGHT = 20 * MESSAGE_CRISPNESS
+
+local FRIENDLINESS_CACHE_COOLDOWN = 1 -- Friendliness is cached periodically while within draw distance.
+
+surface.CreateFont( "CFC_SpawnPoints_3D2DMessage", {
+    font = "Arial",
+    size = MESSAGE_FONT_SIZE,
+    weight = 500,
+    antialias = true,
+    shadow = false,
+} )
+
+
+----- PRIVATE FUNCTIONS -----
+
+local function shouldShowPointSpawnCooldown( spawnPoint, ply, now )
+    if now >= spawnPoint:GetCreationCooldownEndTime() then return false end
+    if hook.Run( "CFC_SpawnPoints_IgnorePointSpawnCooldown", spawnPoint, ply ) then return false end
+
+    return true
+end
+
+local function shouldShowPlayerSpawnCooldown( _spawnPoint, ply, now )
+    if now >= CFC_SpawnPoints.GetSpawnCooldownEndTime( ply ) then return false end
+    if hook.Run( "CFC_SpawnPoints_IgnorePlayerSpawnCooldown", ply ) then return false end
+
+    return true
+end
+
+
+----- ENTITY METHODS -----
 
 function ENT:Initialize()
+    self._inDrawDistance = false
+    self._isFriendlyCache = false
+    self._nextFriendlinessCacheTime = 0
 end
 
 function ENT:Draw()
-    self:DrawModel()
+    local myTbl = entMeta.GetTable( self )
+    entMeta.DrawModel( self )
+    myTbl.TryDrawMessage( self, myTbl )
+end
+
+function ENT:TryDrawMessage( myTbl )
+    local ply = LocalPlayer()
+    local myPos = entMeta.GetPos( self )
+    local eyeDist = vecMeta.Distance( EyePos(), myPos )
+
+    -- Draw distance check.
+    if eyeDist > MESSAGE_DRAW_DISTANCE then
+        if myTbl._inDrawDistance then -- Leaving draw distance
+            myTbl._inDrawDistance = false
+        end
+
+        return
+    end
+
+    if not self._inDrawDistance then -- Entering draw distance
+        self._inDrawDistance = true
+        self:UpdateFriendlinessCache( true ) -- Force update on enter
+    else
+        self:UpdateFriendlinessCache( false ) -- Periodic update while in draw distance
+    end
+
+    -- Health message
+    local maxHealth = self:GetPointMaxHealth()
+    local isFriendly = self._isFriendlyCache
+    local alpha = 255
+
+    if MESSAGE_FADE_START_DISTANCE ~= 0 and eyeDist > MESSAGE_FADE_START_DISTANCE then
+        local farFrac = ( eyeDist - MESSAGE_FADE_START_DISTANCE ) / ( MESSAGE_DRAW_DISTANCE - MESSAGE_FADE_START_DISTANCE )

```

Comment:
Would love to see a couple of extra vars here - I get the idea, but would need to spend a little longer if I wanted to try changing or modifying it